### DISCUSSION TITLE
Filters - Changing parameter values from client side

### DISCUSSION THREAD

**Discussion by sojan1** [Mimetis/Dotmim.Sync#1081] [Q&A] [answered]
Hi 
I want the sync to only synchronise the current parameters set in the client. 
how can i change the parameter? 

link i am refering to [https://dotmimsync.readthedocs.io/Filters.html]

**Client side** 
            var parameters = new SyncParameters
                {
                    { "CLNT_ID", "D4" } 
                };

**if i want to change D4 With D5? **i.e.** i dont want to synchronise D4.**. how can i do it?

**Server side** 
            var setup = new SyncSetup("_clnt", "PROJECT");

            var clntFilter = new SetupFilter("_clnt");
            clntFilter.AddParameter("CLNT_ID", "_clnt", true);
            clntFilter.AddWhere("CLNT_ID", "_clnt", "CLNT_ID");
            setup.Filters.Add(clntFilter);

            var projFilter = new SetupFilter("PROJECT");
            projFilter.AddParameter("CLNT_ID", "PROJECT", true);
            projFilter.AddWhere("CLNT_ID", "PROJECT", "CLNT_ID");
            setup.Filters.Add(projFilter);

            builder.Services.AddSyncServer<SqlSyncProvider>(connectionString, setup, options);

Thanks

**Comment by Mimetis** [comment]
You can change D4 to D5

**Reply by sojan1** [reply]
(**Client to Hub**)
 What i have noticed is, When synchronising data with D5 **after D4** on the same client device, it synchronises both D4 and D5, even when my parameter is D5.  This will occurs only **if the device's data with parameter D4 gets updated** before syncing, even if the parameter to synchronize is D5. 

Is there a way to synchronize data based only on client device's current sync-parameters, and not synchronize based on all parameters including previous? 

(Scenerio - The user might have updated data with D4 and D5, but the user only want to synchronize D5 now, for the reason D4 data eventhough updated is not at a state to synchronize with the server yet. )


**Table : scope_info_client** - how can i **synchronize data to the hub based on latest sync_scope_parameter** and not all sync_scope_parameters. Or how can i  **replace the sync_scope_parameter** which will also **stop parameter list from growing** on client.
![image](https://github.com/Mimetis/Dotmim.Sync/assets/111710684/87c3b5d0-f003-4043-9341-48f129b55e09)

**Reply by Mimetis** [reply]
are you using SQLite on client side ?
If yes, `SQLIte` is not applying any filter locally, even if they are defined, since SQLite is only a client database, and filter are applied on server side.

Maybe this is something we need to reconsider

**Reply by Mimetis** [reply]
I remember we had this conversation with @gentledepp, but I dont remember exactly why we changed this behavior

**Reply by sojan1** [reply]
Yes, i am using SQLite on client side.

**Reply by Mimetis** [reply, marked as answer]
I've published a new beta version [1.0.0-beta-1514](https://www.nuget.org/packages/Dotmim.Sync.Core/1.0.0-beta-1514)
You should try it, I've added the filter on the sqlite side

**Reply by sojan1** [reply]
Hi,  

**It works very well. :)** Thank you very much.

Can i ask you few more questions.

**1.** Is it possible to add a **new CLNT_ID**  (eg: D7)  **on client's SQLite** and have that D7 added in the filter parameters to push it to the Server. Or how can i add a new record(new clnt_id) on client side and synchronize it with the server.

**2.** is there a way to have **more than one value**(**D4, D6**) filtered **against same field (CLNT_ID )** . ? **Will multiple parameters work if i put something like below ?**

**Server Side-** 
            var clntFilter = new SetupFilter("_clnt");
            clntFilter.AddParameter("CLNT_ID", "_clnt", true);
            clntFilter.**AddCustomWhere**("{{{CLNT_ID}}} **IN (@CLNT_ID)**");
            setup.Filters.Add(clntFilter);

**Client Side-** 
        var parameters = new SyncParameters
                {
                    { "CLNT_ID",(  "D4, D6"  )}
                };
 
 


**3.** What is your opinion on do the provisioning using a windows application on a server machine, and only perform the synchonisation using the web api http://localhost:81/api/sync. 
(meaning not to trigger a provisioning job using client device, instead do the provisioning always on server side using a windows form or console application.)

Thanks

**Reply by Mimetis** [reply]
### 1)  Adding a new parameter "on the fly"

Yes it's possible. It's an advanced scenario, so be sure you understand what you are doing here.
Here is small piece of code as demo:

``` cs
// classic sync for A_BIKES
var paramsABikes = new SyncParameters { { "ProductCategoryId", "A_BIKES" } };

// Sync server. Create the local schema, Get all the A_BIKES rows, and apply them locally
var s1 = await agent.SynchronizeAsync(setup, paramsABikes, progress);
Console.WriteLine(s1);

// Adding a R_BIKES product category, and we want to upload it, despite the fact we don't have yet a correct scope
await DBHelper.AddProductCategoryRowAsync(agent.LocalOrchestrator.Provider,  "R_BIKES", default, "R Bikes category");

// the params to use                
var paramsRBikes = new SyncParameters { { "ProductCategoryId", "R_BIKES" } };

// You need to do this trick only ONE TIME.
// Once it's done, you don't have to do it on every sync, of course
// -----------------------------------------------------------------
// Get the scope client for R_BIKES from local storage
// The GetScopeInfoClientAsync method will create the scope if it does not exists in the local database
var scopeClientRBikes = await agent.LocalOrchestrator.GetScopeInfoClientAsync(syncParameters: paramsRBikes);
// hack the last sync datetime
scopeClientRBikes.LastSyncTimestamp = 0;
scopeClientRBikes.LastSync = DateTime.Now.AddDays(-1);
// Save it
await agent.LocalOrchestrator.SaveScopeInfoClientAsync(scopeClientRBikes);
// -----------------------------------------------------------------

// now sync
var s2 = await agent.SynchronizeAsync(setup, paramsRBikes, progress);
Console.WriteLine(s2);
```
### 2 AddCustomWhere()

`AddCustomWhere` is your best bet, I guess. 
Be careful on using it, and check the documentation about [AddCustomWhere](https://dotmimsync.readthedocs.io/Filters.html#the-addcustomwhere-method).

Especially, be careful about `sync_row_is_tombstone`

Be careful also to use a SQL statement that will work on both client & server databases (`SQL Server` / `SQLite`)

Honestly, I would prefer to go with several parameters and one sync per parameter, and not using the `AddCustomWhere` method.

At the end, you will have **N** sync, for **N** parameters, but you will be more in control.
Just my opinion

### 3 Provisionining

If yo have access to the client through a TCP connection for the provision step, it will be way faster, and it's a good thing.

**Comment by sojan1** [comment]
1. Adding a new parameters on the fly- 
               **Will 2 scopes solve the problem one with filters and one without filters just for the main tables**?

**Reply by Mimetis** [reply]
Probably, I think you should test it :)

**Comment by sojan1** [comment]
**It worked** the way you proposed in the above post :) - **Adding a new parameter "on the fly"**.  

Only thing is it is adding an extra scope on the client's **Scope_Info_Client table** with the name **DefaultScope**

**Comment by sojan1** [comment]
Hi @Mimetis,  Does the latest version **[1.0.0-beta-1517]** have all the changes made in **1.0.0-beta-1514** ?

Thanks,

**Reply by Mimetis** [reply]
yes
